lavadecode.1
'\" t
'\"
'\"	DO NOT EDIT! This file is generated from lavadecode.1in
'\"
.TH lavadecode 1 "Sun Apr 25 16:22:30 2021" "lavadecode 0.0"
'\"
'\"==========================================================================
'\"	STRINGS and MACROS
'\"==========================================================================
'\"
'\"	Define strings for special characters that nroff doesn't have
'\"
'\"	N.B. using .if cua to test for special characters did not work.
'\"
.if !'\*[.T]'ps' .ds ua up
.if  '\*[.T]'ps' .ds ua \(ua
.if !'\*[.T]'ps' .ds da down
.if  '\*[.T]'ps' .ds da \(da
.if !'\*[.T]'ps' .ds <- left
.if  '\*[.T]'ps' .ds <- \(<-
.if !'\*[.T]'ps' .ds -> right
.if  '\*[.T]'ps' .ds -> \(->
'\"
'\"     bx - box a Courier string for making keycaps
'\"
'\"	N.B. this mess is to make the line drawing come out only
'\"	if we are really generating postscript
'\"
.de bx
.ie !'\*[.T]'ps' \{\
.	RB [ \\$1 ]\\$2
.\}
.el \{\
.	ie !r ps4html \{\
'\" \(br\|\s-1\f(CB\\$1\fP\s+1\|\(br\l'|0\(rn'\l'|0\(ul'
.		ft CW
.		nr par*bxw \w'\\$1'+.4m 
\Z'\v'.25m'\D'l 0 -1m'\D'l \\n[par*bxw]u 0'\D'l 0 1m'\D'l -\\n[par*bxw]u 0''\
\Z'\h'.2m'\s-1\\$1\s+1'\
\h'\\n[par*bxw]u'\\$2
.		ft P
.	\}
.	el \{\
.		RB [ \\$1 ]\\$2
.	\}
.\}
..
'\"
'\" strings to set current color (null with old groff)
'\"
.if mred .ds red \m[red]
.if mgreen .ds green \m[green]
.if mblue .ds blue \m[blue]
.if mblack .ds black \m[black]
.if mblack .ds mP \mP
'\"
'\" fix for grotty + xterm. We call for orange, grotty outputs yellow,
'\" but xterm displays yellow as orange.  The cycle is complete.
'\"
.if n .defcolor orange rgb #ffff00
'\"
'\" color <color> - set the current color (ignores request with old groff)
'\"
.de color
.if mred \m[\\$1]\c
..
'\"
'\" colorword <color> <word> - colorize a word (ignored by old groff)
'\"
.de colorword
.ie m\\$1 \m[\\$1]\\$2\mP\c
.el \\$2\c
..
'\"
'\" colbox <fg> <bg> <word> - colorize a word in a filled box
'\"
.de colbox
.ie mred \M[\\$2]\
\v'+.167v'\
\D'P 0 -0.9v  \w'\\$3'u 0  0 +0.9v   -\w'\\$3'u 0'\
\v'-.167v'\
\m[\\$1]\\$3\mP\MP
.el \\$3\c
..
'\"
'\"	Macros for doing pdfmarks
'\"
.de specialps
.if  '\*[.T]'ps' \\k_\X'ps: \\$*'\h'|\\n_u'\c
..
'\"
'\" pdfmark PDFMARKCODE
'\"
.ds pdfmarks
.if d pdfmarks \{\
.de pdfmark
.	specialps exec [\\$1 pdfmark
..
'\"
'\" pdfdest LINKNAME
'\"
.de pdfdest
.pdfmark "/Dest /\\$1 /View [/XYZ -5 PL null] /DEST"
..
'\"
'\" pdfbookmark COUNT LINKNAME STRING
'\"
.de pdfbookmark
.   pdfmark "/View [/XYZ 44 730 1.0] /Count \\$1 /Dest /\\$2 /Title (\\$3) /OUT"
..
'\"
'\"	Define the SH and SS macros to save pdfmark information
'\"	in "arrays" of numbers and strings.
'\"
.if !r rr_n \{\
.nr rr_n 0 1
.am SH
.	nr rr_levels!\\n+[rr_n] 2
.	ds rr_labels!\\n[rr_n] \\$*
.	pdfdest Link\\n[rr_n]
..
.am SS
.	nr rr_levels!\\n+[rr_n] 3
.	ds rr_labels!\\n[rr_n] \\$*
.	pdfdest Link\\n[rr_n]
..
.\}
'\"
'\"	Called at the end of the document to generate the pdfmark outline
'\"
.de pdf_outline
.nr rr_levels!\\n+[rr_n] 1
.nr rr_i 0 1
.while \\n+[rr_i]<\\n[rr_n] \{\
.   nr rr_ip1 \\n[rr_i]+1
.   nr rr_count 0
.   if \\n[rr_levels!\\n[rr_ip1]]>\\n[rr_levels!\\n[rr_i]] \{\
.       nr rr_j \\n[rr_i] 1
.       while \\n+[rr_j]<\\n[rr_n] \{\
.           if \\n[rr_levels!\\n[rr_j]]<=\\n[rr_levels!\\n[rr_i]] \{\
.               break
.           \}
.           if \\n[rr_levels!\\n[rr_j]]==(\\n[rr_levels!\\n[rr_i]]+1) \{\
.               nr rr_count \\n[rr_count]+1
.           \}
.       \}
.   \}
.   ds hhh \\*[rr_labels!\\n[rr_i]]
.   pdfbookmark -\\n[rr_count] Link\\n[rr_i] "\\*[hhh]"
.\}
..
'\"
'\" Some postscript to make pdfmarks harmless on old interpreters...
'\"
.specialps "def /pdfmark where {pop} {userdict /pdfmark /cleartomark load put} ifelse"
'\"
'\" Force display of Bookmarks in Acrobat when document is viewed.
'\"
.pdfmark "[/PageMode /UseOutlines /Page 1 /View [/XYZ null null null] /DOCVIEW"
'\"
'\" Output the document info in pdfmarks
'\"
.pdfmark "\
	/Title (\*[an-title](\*[an-section])) \
	/Subject (\*[an-title] Manual Page) \
	/Author (Rick Richardson) \
	/Keywords (printing) \
	/Creator (groff \n(.x.\n(.y.\n(.Y -man) \
	/CreationDate (\*[an-extra1]) \
	/ModDate (\*[an-extra1]) \
	/DOCINFO"
\}
'\"
'\" The manual page name is only 1st level mark
'\"
.nr rr_levels!\n+[rr_n] 1
.ds rr_labels!\n[rr_n] \*[an-title](\*[an-section])
.pdfdest Link\n[rr_n]
'\"
'\"==========================================================================
'\"	MANUAL PAGE SOURCE
'\"==========================================================================
.SH NAME
lavadecode \- Decode a LAVAFLOW stream into human readable form.
.SH SYNOPSIS
.B lavadecode
.RI [ options "] <" lavaflow-file
.SH DESCRIPTION
.B lavadecode
decodes a LAVAFLOW stream into human readable form.
.P
A LAVAFLOW stream is the printer language used by some Konica Minolta
printers, such as the KM magicolor 2530 DL.

.SH COMMAND LINE OPTIONS
These are the options that can appear on the command line.
.TP
.BI \-d\0 basename
Basename of .pbm file for saving decompressed planes.
.TP
.BI \-h
Print hex file offsets.
.TP
.BI \-o
Print file offsets.
.TP
.BI \-D\0 level
Set Debug level [0].

.SH EXAMPLES
Decode an LAVAFLOW stream file created by foo2lava.

.nf
.ft CW
$ lavadecode -h < testpage.prn
     0: \\033%-12345X@PJL JOB NAME="stdin"
    1f: \\033%-12345X@PJL JOB USERNAME=""
    3d: \\033%-12345X@PJL JOB TIMESTAMP="07/20/2007"
    66: \\033%-12345X@PJL JOB OSINFO="Linux/2.6.20-1.2316.fc5"
    99: \\033%-12345X@PJL ENTER LANGUAGE=LAVAFLOW
    bf: \\033E                  RESET
    c1: \\033&l0S               DUPLEX: [off]
    c6: \\033&l0G
    cb: \\033&u1200D            X RESOLUTION: [1200]
    d3: \\033&l1X               COPIES: [1]
    d8: \\033&x1X               TRANSMIT ONCE COPIES: [1]
    dd: \\033&l0O               ORIENTATION: [port]
    e2: \\033*r1U               NBIE: [1]
    e7: \\033*g8W               BW/COLOR: [8]
                                fmt=2 np=1
                                BLACK:  X=1200, Y=600, unk=0, #=4(2)
    f4: \\033*b1234M            COMPRESSION: [1234] (JBIG)
    fc: \\033&l2A               PAGE SIZE: [letter]
   101: \\033&l255H             PAPER SOURCE: [auto]
   108: \\033&l0M               MEDIA TYPE: [plain]
   10d: \\033&l0E               TOP MARGIN: [0]
   112: \\033*r9792S            X RASTER: [9792,0x2640]
   11a: \\033*r6400T            Y RASTER: [6400,0x1900]
   122: \\033&l0U
   127: \\033&l0Z
   12c: \\033*p200X             X OFFSET: [200]
   133: \\033*p200Y             Y OFFSET: [200]
   13a: \\033*r1A               [Page 1]
   13f: \\033*b20V              [black]
                                DL = 0, D = 0, P = 1, - = 0, XY = 9792 x 6400
                                L0 = 128, MX = 0, MY = 0
                                Order   = 3  ILEAVE SMID
                                Options = 92  LRLTWO TPDON TPBON DPON
                                50 stripes, 0 layers, 1 planes
   159: \\033*b65536V           JBIG data (first) [65536,0x10000]
                         ff 02 c2 79 54 3e be e1 a0 de 08 9a b1 d2 c2 59
                        ... ae 88 ef a7 c7 96 d3 96 a6 d7 2c 06 38 75 22 44
 10162: \\033*b26432W           JBIG data (end) [26432,0x6740]
                         0e 89 66 ce 01 41 41 41 41 41 41 41 41 41 41 41
                        ... 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
 168ab: \\033*x3887138K         BLACK DOTS: [3887138]
 168b6: \\033*x58781662W        BLACK WHITEDOTS: [58781662]
 168c2: \\033*rC                END PAGE
 168c6: \\033&l0H               PAPER SOURCE: [eject]
 168cb: \\033E                  RESET
 168cd: \\033%-12345X
.ft P
.fi

.SH FILES
.BR /usr/bin/lavadecode
.SH SEE ALSO
.BR foo2lava-wrapper (1),
.BR foo2lava (1)
.SH "AUTHOR"
Rick Richardson <rick.richardson@comcast.net>
.br
http://foo2lava.rkkda.com/
'\"
'\"
'\"
.em pdf_outline


ChangeLog
2020-06-10	Rick Richardson <rick.richardson@comcast.net>
	* osx-hotplug: Add PREFIX=/usr or PREFIX=/usr/local to cc rules

2020-05-05:	Rick Richardson <rick.richardson@comcast.net>
	* foo2hbpl2.c: SaveToner implemented for color pages,
	  thanks to bedi@3mtech.sk .

2020-04-26:	Rick Richardson <rick.richardson@comcast.net>
	* foo2hbpl2.c: Bug with SaveToner

2020-02-07:	Rick Richardson <rick.richardson@comcast.net>
	* Add Cassette 2 to foo2oak-z1-InputSlot.xml (KM-1635)

2019-09-09:	Rick Richardson <rick.richardson@comcast.net>
	* Add SmartInstallDisable-Tool.run since it was gone from
	      http://hplipopensource.com/hplip-web/smartinstall/
	  See HP SMART INSTALL NOTES in the INSTALL doc.

2019-04-13:	Rick Richardson <rick.richardson@comcast.net>
	* Makefile: Add install-aa to disable AppArmor on SUSE

2018-09-10:	Rick Richardson <rick.richardson@comcast.net>
	* Makefile: target ppd:	revert the 09-06 change.

2018-09-06:	Rick Richardson <rick.richardson@comcast.net>
	* Makefile: target ppd:
	    Use the older foomatic-ppdfile which only has postscript,
	    not postscript and pdf, for making ppd's.

2017-12-02	Rick Richardson <rick.richardson@comcast.net>
	* New Printers: HP LaserJet Pro M12a, M12w

2017-03-20	Rick Richardson <rick.richardson@comcast.net>
	* Add ddstdecode, for decoding some Ricoh printers.

2016-10-21	Rick Richardson <rick.richardson@comcast.net>
	* Makefile: install-hotplug-prog
	    Add /lib/systemd/systemd-udevd to the list for Ubuntu 16.

2016-09-02	Rick Richardson <rick.richardson@comcast.net>
	* New Printer: Xerox Phaser 3040 using the foo2hbpl2 driver.

2016-08-30	Rick Richardson <rick.richardson@comcast.net>
	* New Printer: Xerox Phaser 3010 using the foo2hbpl2 driver.

2016-03-13	Rick Richardson <rick.richardson@comcast.net>
	* icc2ps/lcms.h: USE_CUSTOM_SWAB defined.

2015-10-24	Rick Richardson <rick.richardson@comcast.net>
	* foomatic-db/printer/Epson-AcuLaser_CX17NF.xml:
		Change margins on Epson AcuLaser CX17NF.

2015-10-11	Rick Richardson <rick.richardson@comcast.net>
	* New Printer: Epson AcuLaser CX17NF using the foo2hbpl2 driver.

2015-07-04	pavel.shvagirev	<pavel.shvagirev@gmail.com>
	* Modify: hplj10XX to not call us twice on Ubuntu 14.

2015-05-11	Rick Richardson <rick.richardson@comcast.net>
	* Bug in foomatic-db/printer/Xerox-Phaser_6115MFP.xml . Should use
	  foo2lava driver. Thanks to user "ophers".

2015-02-11	Rick Richardson <rick.richardson@comcast.net>
	* New Printer: Oki C511 using foo2hiperc.

2014-12-21	Rick Richardson <rick.richardson@comcast.net>
	* osx-hotplug: make it work with OS X 10.10

2014-12-04	Rick Richardson <rick.richardson@comcast.net>
	* make cups: add "Sandboxing Relaxed" to cups-files.conf iff cups >= 2

2014-09-25	Rick Richardson <rick.richardson@comcast.net>
	* foo2hblp2, foo2hblp2-wrapper, foomatic et al:
	    Add paper size A5.
	* foo2hblp2 foomatic db entries:
	    Changed resolution to 600x600, 1200x600, and 1200x1200.
	    Note: need to recreate the printers if you want more choices!

2014-05-19	Rick Richardson <rick.richardson@comcast.net>
	* HP CLJ CP1215:  Adjust margins: *HWMargins: 18 14.40 18 14

2014-05-04	Rick Richardson <rick.richardson@comcast.net>
	* foo2hbpl2-wrapper: Change PAPER=21 to PAPER=0 for custom page sizes.

2014-04-17	Rick Richardson <rick.richardson@comcast.net>
	* New Printer: Fuji Xerox DocuPrint CM215 MFP using the foo2hbpl2
	driver.

2014-04-08	Rick Richardson <rick.richardson@comcast.net>
	* Upgrade to JBIGKIT 2.1

2014-04-01	Rick Richardson <rick.richardson@comcast.net>
	* foo2hbpl2 foomatic-db files and PPDs:
	    Bug with default PageSize.
	    PLEASE delete and recreate the foo2hbpl2 printers.

2014-03-22	Rick Richardson <rick.richardson@comcast.net>
	* hbpldecode: another redundancy from Dave Coffin

2014-03-18	Rick Richardson <rick.richardson@comcast.net>
	* hbpldecode: More envelope sizes.

2014-03-17	Rick Richardson <rick.richardson@comcast.net>
	* foo2hbpl2.c: Bug, was using A4(1) instead of letter(4) by default
	* hbpldecode: Print out paper codes.

2014-03-15	Rick Richardson <rick.richardson@comcast.net>
	* hbpldecode: More changes from Dave.  Still gets 

lavadecode.c
/*
 * $Id: lavadecode.c,v 1.36 2014/06/27 14:41:37 rick Exp $
 */

/*b
 * Copyright (C) 2003-2006  Rick Richardson
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 * Author: Rick Richardson <rick.richardson@comcast.net>
b*/

#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <string.h>
#include <unistd.h>
#include <errno.h>

#include "jbig.h"

/*
 * Global option flags
 */
int	Debug = 0;
char	*DecFile;
int	PrintOffset = 0;
int	PrintHexOffset = 0;

void
debug(int level, char *fmt, ...)
{
    va_list ap;

    if (Debug < level)
	return;

    va_start(ap, fmt);
    vfprintf(stderr, fmt, ap);
    va_end(ap);
}

int
error(int fatal, char *fmt, ...)
{
	va_list ap;

	fprintf(stderr, fatal ? "Error: " : "Warning: ");
	if (errno)
	    fprintf(stderr, "%s: ", strerror(errno));
	va_start(ap, fmt);
	vfprintf(stderr, fmt, ap);
	va_end(ap);
	if (fatal > 0)
	    exit(fatal);
	else
	{
	    errno = 0;
	    return (fatal);
	}
}

void
usage(void)
{
    fprintf(stderr,
"Usage:\n"
"	lavadecode [options] < zjs-file\n"
"\n"
"	Decode a LAVAFLOW stream into human readable form.\n"
"\n"
"	A LAVAFLOW stream is the printer language used by some Konica\n"
"	Minolta printers, such as the magicolor 2530 DL and 2490 MF.\n"
"\n"
"\n"
"Options:\n"
"       -d basename Basename of .pbm file for saving decompressed planes\n"
"       -o          Print file offsets\n"
"       -h          Print hex file offsets\n"
"       -D lvl      Set Debug level [%d]\n"
    , Debug
    );

    exit(1);
}

void
print_bih(unsigned char bih[20])
{
    unsigned int xd, yd, l0;
    char hdr[] = "\n\t\t\t\t";

    if (!PrintOffset && !PrintHexOffset) hdr[strlen(hdr)-1] = 0;

    xd = (bih[4] << 24) | (bih[5] << 16) | (bih[6] << 8) | (bih[7] << 0);
    yd = (bih[8] << 24) | (bih[9] << 16) | (bih[10] << 8) | (bih[11] << 0);
    l0 = (bih[12] << 24) | (bih[13] << 16) | (bih[14] << 8) | (bih[15] << 0);

    printf("%sDL = %d, D = %d, P = %d, - = %d, XY = %d x %d",
	 hdr, bih[0], bih[1], bih[2], bih[3], xd, yd);

    printf("%sL0 = %d, MX = %d, MY = %d",
	 hdr, l0, bih[16], bih[17]);

    printf("%sOrder   = %d %s%s%s%s%s", hdr, bih[18],
	bih[18] & JBG_HITOLO ? " HITOLO" : "",
	bih[18] & JBG_SEQ ? " SEQ" : "",
	bih[18] & JBG_ILEAVE ? " ILEAVE" : "",
	bih[18] & JBG_SMID ? " SMID" : "",
	bih[18] & 0xf0 ? " other" : "");

    printf("%sOptions = %d %s%s%s%s%s%s%s%s", hdr, bih[19],
	bih[19] & JBG_LRLTWO ? " LRLTWO" : "",
	bih[19] & JBG_VLENGTH ? " VLENGTH" : "",
	bih[19] & JBG_TPDON ? " TPDON" : "",
	bih[19] & JBG_TPBON ? " TPBON" : "",
	bih[19] & JBG_DPON ? " DPON" : "",
	bih[19] & JBG_DPPRIV ? " DPPRIV" : "",
	bih[19] & JBG_DPLAST ? " DPLAST" : "",
	bih[19] & 0x80 ? " other" : "");
    printf("%s%u stripes, %d layers, %d planes",
	hdr,
	((yd >> bih[1]) +  ((((1UL << bih[1]) - 1) & xd) != 0) + l0 - 1) / l0,
	bih[1] - bih[0], bih[2]);
}

void
print_config(unsigned char *c)
{
    char hdr[] = "\n\t\t\t\t";

    if (!PrintOffset && !PrintHexOffset) hdr[strlen(hdr)-1] = 0;

    if (c[1] == 1)
    {
	printf("%sfmt=%d np=%d",
	    hdr, c[0], c[1]);
	printf("%sBLACK:	X=%d, Y=%d, unk=%d, #=%d(%d)",
	    hdr, (c[2]<<8) + c[3], (c[4]<<8) + c[5], c[6], 1 << c[7], c[7]);
    }
    else if (c[1] == 4)
    {
	printf("%sfmt=%d np=%d",
	    hdr, c[0], c[1]);
	printf("%sYEL:	X=%d, Y=%d, unk=%d, #=%d(%d)",
	    hdr, (c[2]<<8) + c[3], (c[4]<<8) + c[5], c[6], 1 << c[7], c[7]);
	printf("%sMAG:	X=%d, Y=%d, unk=%d, #=%d(%d)",
	    hdr, (c[8]<<8) + c[9], (c[10]<<8) + c[11], c[12], 1 << c[13], c[13]);
	printf("%sCYA:	X=%d, Y=%d, unk=%d, #=%d(%d)",
	    hdr, (c[14]<<8) + c[15], (c[16]<<8) + c[17], c[18], 1 << c[19], c[19]);
	printf("%sBLK:	X=%d, Y=%d, unk=%d, #=%d(%d)",
	    hdr, (c[20]<<8) + c[21], (c[22]<<8) + c[23], c[24], 1 << c[25], c[25]);
    }
    else
	error(1, "config image data is not 8 or 26 bytes!\n");
}

void
proff(int curOff)
{
    if (PrintOffset)
	printf("%d:	", curOff);
    else if (PrintHexOffset)
	printf("%6x:	", curOff);
}

void
decode(FILE *fp)
{
    int		c;
    int		rc;
    FILE	*dfp = NULL;
    int		pageNum = 0;
    int		curOff = 0;
    struct jbg_dec_state	s[5];
    unsigned char	bih[20];
    int			bihlen = 0;
    int			imageCnt[5] = {0,0,0,0,0};
    int         	pn = 0;
    int			nbie = 0;
    int         	incrY = 0;
    int			totSize = 0;
    char		buf[1024];
    char		*strpage[837+1];
    int			i;
    int			totval = 0;

    for (i = 0; i < sizeof(strpage)/sizeof(strpage[0]); ++i)
	strpage[i] = "unk";
    strpage[1] = "exec";
    strpage[2] = "letter";
    strpage[3] = "legal";
    strpage[25] = "a5";
    strpage[26] = "a4";
    strpage[45] = "b5jis";
    strpage[65] = "b5iso";
    strpage[80] = "envMonarch";
    strpage[81] = "env#10";
    strpage[90] = "envDL";
    strpage[91] = "envC5";
    strpage[92] = "envC6";
    strpage[101] = "custom";
    strpage[835] = "photo4x6";
    strpage[837] = "photo10x15";

    while (fgets(buf, sizeof(buf), fp))
    {
	proff(curOff);
	if (buf[0] == '\033')
	{
	    printf("\\033");
	    fputs(buf+1, stdout);
	}
	else
	    fputs(buf, stdout);
	curOff += strlen(buf);

	if (strcmp(buf, "\033%-12345X@PJL ENTER LANGUAGE=LAVAFLOW\n") == 0
	    || strcmp(buf, "@PJL ENTER LANGUAGE=LAVAFLOW\n") == 0)
	{
	    int		state = 0;
	    char	intro = 0, groupc = 0;
	    int		neg = 0, val = 0, pres = 0;
	    size_t		cnt;
	    unsigned char	ch;
	
	    while ( (c = fgetc(fp)) != EOF)
	    {
		curOff++;
		switch (state)
		{
		case 0:
		    if (c == '\033')
		    {
			state = '\033';
			proff(curOff-1);
		    }
		    else
		    {
			proff(curOff-1);
			printf("ch=0x%x (%s)\n", c, c==0x0c ? "form feed" : "");
		    }
		    break;
		case '\033':
		    if (c >= 'A' && c <= 'Z')
		    {
			printf("\\033%c", c);
			if (c == 'E')
			    printf("\t\t\tRESET");
			printf("\n");
			state = 0;
		    }
		    else
		    {
			intro = c;
			state = 'i';
		    }
		    break;
		case 'i':
		    if (c >= 'A' && c <= 'Z')
			state = 0;
		    else
		    {
			groupc = c;
			state = 'v';
			pres = neg = val = 0;
		    }
		    break;
		case 'v':
		    if (c >= 'A' && c <= 'Z')
		    {
			#define STRARY(X, A) \
			    ((X) >= 0 && (X) < sizeof(A)/sizeof(A[0])) \
			    ? A[X] : "UNK"
			char *strduplex[] = {
			    "off", "long", "short", "manlong", "manshort"
			    };
			char *strorient[] = { "port", "land" };
			char *strsource[] = {
			    /*00*/ "eject", "tray1", "unk", "unk", "tray2",
			    /*05*/ "unk", "unk", "auto"
			    };
			char *strmedia[] = { 
			    /*00*/ "plain", "unk", "unk", "unk", "transparency",
			    /*05*/ "unk", "unk", "unk", "unk", "unk",
			    /*10*/ "unk", "unk", "unk", "unk", "unk",
			    /*15*/ "unk", "unk", "unk", "unk", "unk",
			    /*20*/ "thick", "unk", "env", "letterhead", "unk",
			    /*25*/ "postcard", "labels", "recycled", "glossy",
			    };

			if (neg) val = -val;

			if (pres)
			    printf("\\033%c%c%d%c", intro, groupc, val, c);
			else
			    printf("\\033%c%c%c\t", intro, groupc, c);
			state = 0;
			if (intro == '&' && groupc == 'l' && c == 'S')
			    printf("\t\tDUPLEX: [%s]", STRARY(val, strduplex));
			if (intro == '&' && groupc == 'u' && c == 'D')
			    printf("\t\tX RESOLUTION: [%d]", val);
			if (intro == '&' && groupc == 'l' && c == 'X')
			    printf("\t\tCOPIES: [%d]", val);
			if (intro == '&' && groupc == 'x' && c == 'X')
			    printf("\t\tTRANSMIT ONCE COPIES: [%d]", val);
			if (intro == '&' && groupc == 'l' && c == 'O')
			    printf("\t\tORIENTATION: [%s]",
				STRARY(val, strorient));
			if (intro == '&' && groupc == 'l' && c == 'A')
			    printf("\t\tPAGE SIZE: [%s]",
				STRARY(val, strpage));
			if (intro == '&' && groupc == 'l' && c == 'E')
			    printf("\t\tTOP MARGIN: [%d]", val);
			if (intro == '&' && groupc == 'l' && c == 'H')
			    printf("\t\tPAPER SOURCE: [%s]",
				STRARY(val&7, strsource));
			if (intro == '&' && groupc == 'l' && c == 'M')
			    printf("\t\tMEDIA TYPE: [%s]",
				STRARY(val, strmedia));
			if (intro == '*' && groupc == 'r' && c == 'S')
			    printf("\t\tX RASTER: [%d,0x%x]", val, val);
			if (intro == '*' && groupc == 'r' && c == 'T')
			    printf("\t\tY RASTER: [%d,0x%x]", val, val);
			if (intro == '&' && groupc == 'f' && c == 'F')
			    printf("\t\tX CUSTOM: [%d,0x%x]", val, val);
			if (intro == '&' && groupc == 'f' && c == 'G')
			    printf("\t\tY CUSTOM: [%d,0x%x]", val, val);
			if (intro == '*' && groupc == 'r' && c == 'U')
			{
			    nbie = val & 7;
			    printf("\t\tNBIE: [%d]", nbie);
			}
			if (intro == '*' && groupc == 'g' && c == 'W')
			{
			    unsigned char config[1024];
			
			    printf("\t\tBW/COLOR: [%d]", val);
			    rc = fread(config, val, 1, fp);
			    curOff += val;
			    print_config(config);
			}
			if (intro == '*' && groupc == 'b' && c == 'M')
			{
			    printf("\t\tCOMPRESSION: [%d]", val);
			    if (val == 1234)
				printf(" (JBIG)");
			    else
				printf(" (unknown compression scheme)");
			}
			if (intro == '*' && groupc == 'r' && c == 'C')
			    printf("\t\tEND PAGE");
			if (intro == '*' && groupc == 'p' && c == 'X')
			    printf("\t\tX OFFSET: [%d]", val);
			if (intro == '*' && groupc == 'p' && c == 'Y')
			    printf("\t\tY OFFSET: [%d]", val);
			if (intro == '*' && groupc == 'x' && c == 'Y')
			    printf("\t\tYELLOW DOTS: [%d]", val);
			if (intro == '*' && groupc == 'x' && c == 'M')
			    printf("\t\tMAGENTA DOTS: [%d]", val);
			if (intro == '*' && groupc == 'x' && c == 'C')
			    printf("\t\tCYAN DOTS: [%d]", val);
			if (intro == '*' && groupc == 'x' && c == 'K')
			    printf("\t\tBLACK DOTS: [%d]", val);
			if (intro == '*' && groupc == 'x' && c == 'U')
			    printf("\t\tYELLOW WHITEDOTS: [%d]", val);
			if (intro == '*' && groupc == 'x' && c == 'V')
			    printf("\t\tMAGENTA WHITEDOTS: [%d]", val);
			if (intro == '*' && groupc == 'x' && c == 'Z')
			    printf("\t\tCYAN WHITEDOTS: [%d]", val);
			if (intro == '*' && groupc == 'x' && c == 'W')
			    printf("\t\tBLACK WHITEDOTS: [%d]", val);

			if (intro == '%' && val == 12345 && c == 'X')
			    goto out;
			if (intro == '*' && groupc == 'r' && c == 'A')
			{
			    ++pageNum;
			    printf("\t\t[Page %d]", pageNum); 
			    pn = 0;
			}
			if (intro == '*' && groupc == 'b' && c == 'W')
			{
			    printf("\t\tJBIG data (end) [%d,0x%x]", val, val);
			    state = 'd';
			    totval = val;
			}
			if (intro == '*' && groupc == 'b' && c == 'V')
			{
			    if (val == 20)
			    {
				char *color[] =
				    { "yellow", "magenta", "cyan", "black" };
				    
				++pn;
				rc = fread(bih, bihlen = sizeof(bih), 1, fp);
				curOff += bihlen;
				if (nbie == 4 && pn >= 1 && pn <= 4)
				    printf("\t\t[%s]", color[pn-1]);
				else if (nbie == 1)
				    printf("\t\t[%s]", "black");
				else
				    printf("\t\t[%s]", "unknown");
				print_bih(bih);
				if (DecFile)
				{
				    size_t	cnt;

				    jbg_dec_init(&s[pn]);
				    rc = jbg_dec_in(&s[pn], bih, bihlen, &cnt);
				    if (rc == JBG_EIMPL)
					error(1, "JBIG uses unimpl feature\n");
				}
			    }
			    else
			    {
				printf("\t\tJBIG data (first) [%d,0x%x]",
					val, val);
				state = 'd';
				totval = val;
			    }
			}
			printf("\n");
			if (state == 'd')
			    printf("\t\t\t");
		    }
		    else if (c >= '0' && c <= '9')
		    {
			pres = 1;
			val *= 10; val += c - '0';
		    }
		    else if (c == '-')
			neg = 1;
		    else if (c == '+')
			neg = 0;
		    else
			error(1, "c=%d\n", c);
		    break;
		case 'd':
		    --val;
		    #define MAXTOT 16
		    if ((totval-val) <= MAXTOT)
		    {
			printf(" %02x", c);
			if ((totval-val) == MAXTOT)
			    printf("\n\t\t\t...");
		    }
		    else if (val < 16)
			printf(" %02x", c);
		    if (val == 0)
		    {
			state = 0;
			printf("\n");
		    }
		    if (!DecFile || s[pn].s == 0)
			break;

		    ch = c;
		    rc = JBG_EAGAIN;
		    rc = jbg_dec_in(&s[pn], &ch, 1, &cnt);
		    if (rc == JBG_EOK)
		    {
			int	h, w, len;
			unsigned char *image;

			//debug(0, "JBG_EOK: %d\n", pn);
			h = jbg_dec_getheight(&s[pn]);
			w = jbg_dec_getwidth(&s[pn]);
			image = jbg_dec_getimage(&s[pn], 0);
			len = jbg_dec_getsize(&s[pn]);
			if (image)
			{
			    char	buf[512];
			    sprintf(buf, "%s-%02d-%d.pbm",
				    DecFile, pageNum, pn);
			    dfp = fopen(buf, imageCnt[pn] ? "a" : "w");
			    if (dfp)
			    {
				if (imageCnt[pn] == 0)
				    fprintf(dfp, "P4\n%8d %8d\n", w, h);
				imageCnt[pn] += incrY;
				rc = fwrite(image, 1, len, dfp);
				fclose(dfp);
			    }
			}
			else
			    debug(0, "Missing image %dx%d!\n", h, w);
			jbg_dec_free(&s[pn]);
		    }
		    break;
		}
	    }
	out:
	    ;
	}
    }
    if (feof(fp))
	return;

    printf("Total size: %d bytes\n", totSize);
}

int
main(int argc, char *argv[])
{
	extern int	optind;
	extern char	*optarg;
	int		c;

	while ( (c = getopt(argc, argv, "d:hoD:?h")) != EOF)
		switch (c)
		{
		case 'd': DecFile = optarg; break;
		case 'o': PrintOffset = 1; break;
		case 'h': PrintHexOffset = 1; break;
		case 'D': Debug = atoi(optarg); break;
		default: usage(); exit(1);
		}

	argc -= optind;
	argv += optind;

        if (argc > 0)
        {
            FILE        *fp;

            fp = fopen(argv[0], "r");
            if (!fp)
                error(1, "file '%s' doesn't exist\n", argv[0]);
            for (;;)
            {
                decode(fp);
                c = getc(fp); ungetc(c, fp);
                if (feof(fp))
                    break;
            }
            fclose(fp);
        }
        else
        {
	    for(;;)
	    {
		decode(stdin);
		c = getc(stdin); ungetc(c, stdin);
		if (feof(stdin))
		    break;
	    }
	}
	printf("\n");

	exit(0);
}
